### ISSUE TITLE
ext4dist strange timings on ubuntu 16.04

### ISSUE THREAD

**Issue by oity** [iovisor/bcc#790]
on ubuntu 16.04
4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

./ext4dist 5
Tracing ext4 operation latency... Hit Ctrl-C to end.
operation = 'fsync'
               usecs                         : count     distribution
[...]
    4503599627370496 -> 9007199254740991     : 0        |                    |
    9007199254740992 -> 18014398509481983    : 0        |                    |
   18014398509481984 -> 36028797018963967    : 2        |********************|

**Comment by brendangregg** [comment]
Wondering if the ktime bug got backported to older kernels; can someone check if the #728 bug is in 4.4.0-45 ?

**Comment by brendangregg** [comment]
The bug is: timekeeping: Avoid taking lock in NMI path with CONFIG_DEBUG_TIMEKEEPING

The fix is: timekeeping: Fix __ktime_get_fast_ns() regression

Looks like that kernel version has the bug but not the fix. http://changelogs.ubuntu.com/changelogs/pool/main/l/linux/linux_4.4.0-45.66/changelog

@ColinIanKing can you please tell the right person to get the fix integrated (or I can contact someone else at Canonical)?

**Comment by ColinIanKing** [comment]
Thanks for the heads-up, I'll file a bug and we'll get that one fixed.

**Comment by ColinIanKing** [comment]
Just to clarify, I believe the fix being referred to is:

commit 58bfea9532552d422bde7afa207e1a0f08dffa7d
Author: John Stultz <john.stultz@linaro.org>
Date:   Tue Oct 4 19:55:48 2016 -0700
timekeeping: Fix __ktime_get_fast_ns() regression

**Comment by 4ast** [comment]
@ColinIanKing that's correct. It's pretty bad bug that affects not only bpf/bcc, but lttng and perf.

**Comment by ColinIanKing** [comment]
It's cc'd to stable <stable@vger.kernel.org> so our stable updates team should be integrating this fix into the kernel as part of the stable updates process.

**Comment by ColinIanKing** [comment]
Should be in the next Xenial kernel update, it's in the Xenial repo: https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/xenial/commit/?h=master-next&id=a826ba085c671d880ac2b3971b5090e973761846

**Comment by scotte** [comment]
@ColinIanKing any idea when this will be released via xenial-updates? The launchpad bug hasn't been updated. This is a pretty nasty bug that breaks instrumentation. Thanks!

**Comment by ColinIanKing** [comment]
So the fix apparently landed in Xenial updates under a generic updates bug number:

https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1624037

The commit in Xenial is:

commit 8faecee97860f66fdcc67de58803b1fb4e0b0848
Author: John Stultz <john.stultz@linaro.org>
Date:   Tue Aug 23 16:08:21 2016 -0700

    timekeeping: Avoid taking lock in NMI path with CONFIG_DEBUG_TIMEKEEPING

    BugLink: http://bugs.launchpad.net/bugs/1624037

    commit 27727df240c7cc84f2ba6047c6f18d5addfd25ef upstream.

    When I added some extra sanity checking in timekeeping_get_ns() under
    CONFIG_DEBUG_TIMEKEEPING, I missed that the NMI safe __ktime_get_fast_ns()
    method was using timekeeping_get_ns().

    Thus the locking added to the debug checks broke the NMI-safety of
    __ktime_get_fast_ns().

    This patch open-codes the timekeeping_get_ns() logic for
    __ktime_get_fast_ns(), so can avoid any deadlocks in NMI.

    Fixes: 4ca22c2648f9 "timekeeping: Add warnings when overflows or underflows are observed"
    Reported-by: Steven Rostedt <rostedt@goodmis.org>
    Reported-by: Peter Zijlstra <peterz@infradead.org>
    Signed-off-by: John Stultz <john.stultz@linaro.org>
    Link: http://lkml.kernel.org/r/1471993702-29148-2-git-send-email-john.stultz@linaro.org
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

    Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
    Signed-off-by: Kamal Mostafa <kamal@canonical.com>


...so the bug should now be fixed.

**Comment by scotte** [comment]
Great, thanks Colin, appreciate the update!

**Comment by brendangregg** [comment]
BTW, @ColinIanKing, here's a simple test for this bug:

broken system:

```
# perf record -F 9 --cpu 0 --clockid CLOCK_MONOTONIC -- sleep 1
# perf script
         swapper     0 [000] 2942480.468981:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942480.471468:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942480.473959:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942480.475202:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942480.477691:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942481.470226:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942481.472569:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942481.476460:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000] 2942481.478943:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
```

Note the timestamp jump.

fixed (rolled back patch):

```
# perf record -F 9 --cpu 0 --clockid CLOCK_MONOTONIC -- sleep 1
# perf script
         swapper     0 [000]   589.729663:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   589.840772:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   589.951883:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.062994:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.174075:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.285217:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.396328:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.507439:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
         swapper     0 [000]   590.618550:  111111111 cpu-clock:  ffffffff810013aa xen_hypercall_sched_op+0xa (vmlinux)
```

**Comment by scotte** [comment]
Actually, I believe it has not been released yet. The fix is in the changelog for 4.4.0-48.69, but the latest release of the xenial LTS kernel is 4.4.0-47.68 where the bug can still be reproduced.

Specifically:
  * Xenial update to v4.4.25 stable release (LP: #1634153)
    - timekeeping: Fix __ktime_get_fast_ns() regression

**Comment by ColinIanKing** [comment]
4.4.0-51.72 is currently in -proposed so the fix should be released fairly soon now.

**Comment by scotte** [comment]
Confirmed fixed in 4.4.0-51.72

**Comment by goldshtn** [comment]
OK to close then?
